File: campaign_service_record/static/js/i18n.js
/**
 * IL-2 Campaign Tracker - JavaScript i18n Helper
 * 
 * Provides translation functionality for the web frontend (Campaign Service Record).
 * 
 * Usage:
 *     // Initialize (once at startup)
 *     await i18n.init('de');  // or 'en', etc.
 *     
 *     // Translate strings
 *     i18n.t('web.stat.total_kills');  
 *     // → "Gesamt-Abschüsse" (de) or "Total Kills" (en fallback)
 *     
 *     // With parameters
 *     i18n.t('web.title.campaign_detail', {campaign_name: 'Stalingrad'});
 *     // → "Stalingrad - Kampagnen-Dienstakte"
 *     
 *     // Get current locale
 *     const locale = i18n.getLocale();  // → 'de'
 */

const i18n = {
    currentLocale: 'en',
    translations: {},
    initialized: false,
    initInProgress: false,
    _initSequence: 0,
    missingTranslationPlaceholder: '[missing translation]',
    
    /**
     * Initialize i18n with given locale.
     * Loads fallback (en) and requested locale.
     * 
     * @param {string} locale - Locale code ('en', 'de', etc.)
     * @returns {Promise<void>}
     */
    async init(locale = 'en') {
        console.log(`[i18n] Initializing with locale: ${locale}`);
        const initToken = ++this._initSequence;
        this.initInProgress = true;
        
        try {
            // Load fallback (English) first
            this.translations.en = await this.loadLocale('en');
            
            if (!this.translations.en) {
                console.error('[i18n] CRITICAL: Failed to load English fallback!');
            }
            
            // Load requested locale (if different from fallback)
            if (locale !== 'en') {
                this.translations[locale] = await this.loadLocale(locale);
                
                if (!this.translations[locale]) {
                    console.warn(`[i18n] Locale '${locale}' not available, using 'en' fallback`);
                    locale = 'en';
                }
            }
            
            if (initToken !== this._initSequence) {
                console.warn('[i18n] Initialization superseded by a newer request');
                return;
            }

            this.currentLocale = locale;
            this.initialized = true;
            console.log(`[i18n] Initialized successfully with locale: ${locale}`);
        } catch (error) {
            console.error('[i18n] Initialization error:', error);
            if (initToken !== this._initSequence) {
                console.warn('[i18n] Initialization error superseded by a newer request');
                return;
            }
            this.currentLocale = 'en';
            this.initialized = true;
        } finally {
            if (initToken === this._initSequence) {
                this.initInProgress = false;
            }
        }
    },
    
    /**
     * Load translation file for given locale.
     * 
     * @param {string} locale - Locale code
     * @returns {Promise<object>} Translation data
     */
    async loadLocale(locale) {
        try {
            // Try campaign_service_record/static/locales first (symlink)
            let response = await fetch(`/static/locales/${locale}.json`);
            
            // Fallback to root locales directory
            if (!response.ok) {
                response = await fetch(`/locales/${locale}.json`);
            }
            
            if (!response.ok) {
                console.warn(`[i18n] Failed to load locale: ${locale} (${response.status})`);
                return null;
            }
            
            const data = await response.json();
            console.log(`[i18n] Loaded locale: ${locale}`);
            return data;
        } catch (error) {
            console.error(`[i18n] Error loading locale ${locale}:`, error);
            return null;
        }
    },
    
    /**
     * Set current locale.
     * Loads the locale if not already loaded.
     * 
     * @param {string} locale - Locale code
     * @returns {Promise<void>}
     */
    async setLocale(locale) {
        console.log(`[i18n] Setting locale to: ${locale}`);
        
        // Ensure fallback is always loaded
        if (!this.translations.en) {
            this.translations.en = await this.loadLocale('en');
        }
        
        // Load requested locale if not already loaded
        if (!this.translations[locale]) {
            this.translations[locale] = await this.loadLocale(locale);
            
            if (!this.translations[locale]) {
                console.warn(`[i18n] Locale '${locale}' not available, staying with '${this.currentLocale}'`);
                return;
            }
        }
        
        this.currentLocale = locale;
        console.log(`[i18n] Locale set to: ${locale}`);
    },
    
    /**
     * Get current locale code.
     * 
     * @returns {string} Current locale code
     */
    getLocale() {
        return this.currentLocale;
    },
    
    /**
     * Navigate nested object using key parts.
     * 
     * @param {object} data - Object to navigate
     * @param {Array<string>} keyParts - Array of key parts
     * @returns {*} Value at nested location, or null if not found
     */
    _getNestedValue(data, keyParts) {
        let current = data;
        
        for (const part of keyParts) {
            if (current && typeof current === 'object' && part in current) {
                current = current[part];
            } else {
                return null;
            }
        }
        
        return current;
    },

    _isKeyLike(value) {
        return /^[a-z0-9_]+(\.[a-z0-9_]+)+$/i.test(value);
    },

    _resolveKey(key, params = {}) {
        if (!this.initialized) {
            return {
                found: false,
                text: this.missingTranslationPlaceholder,
                localeUsed: this.currentLocale,
                fallbackUsed: false,
                key
            };
        }

        const keyParts = key.split('.');
        let trans = this._getNestedValue(this.translations[this.currentLocale], keyParts);
        let localeUsed = this.currentLocale;
        let fallbackUsed = false;

        if (typeof trans !== 'string' || trans.trim() === '') {
            trans = this._getNestedValue(this.translations.en, keyParts);
            localeUsed = 'en';
            fallbackUsed = true;
        }

        if (typeof trans !== 'string' || trans.trim() === '') {
            return {
                found: false,
                text: this.missingTranslationPlaceholder,
                localeUsed: this.currentLocale,
                fallbackUsed: false,
                key
            };
        }

        let text = trans;
        if (Object.keys(params).length > 0) {
            try {
                text = trans.replace(/\{(\w+)\}/g, (match, param) => {
                    if (param in params) {
                        return params[param];
                    }
                    console.warn(`[i18n] Missing parameter '${param}' for key '${key}'`);
                    return match;
                });
            } catch (error) {
                console.error(`[i18n] Error substituting parameters for key '${key}':`, error);
            }
        }

        return {
            found: true,
            text,
            localeUsed,
            fallbackUsed,
            key
        };
    },

    hasKey(key, locale = null) {
        if (!this.initialized) {
            return false;
        }
        const keyParts = key.split('.');
        const selectedLocale = locale || this.currentLocale;
        const trans = this._getNestedValue(this.translations[selectedLocale], keyParts);
        return typeof trans === 'string' && trans.trim() !== '';
    },
    
    /**
     * Translate key with optional parameters.
     * 
     * Fallback chain: current_locale → 'en' → '[key]'
     * 
     * @param {string} key - Translation key (dot-separated)
     * @param {object} params - Named parameters for string substitution
     * @returns {string} Translated string
     * 
     * @example
     * i18n.t('web.stat.total_kills')
     * // → "Total Kills"
     * 
     * @example
     * i18n.t('web.title.campaign_detail', {campaign_name: 'Stalingrad'})
     * // → "Stalingrad - Campaign Service Record"
     */
    t(key, params = {}) {
        // Ensure initialized
        if (!this.initialized) {
            console.warn('[i18n] Not initialized, using key as-is:', key);
            return this.missingTranslationPlaceholder;
        }

        const resolved = this._resolveKey(key, params);
        if (!resolved.found) {
            console.debug(`[i18n] Translation key not found: ${key}`);
        }
        return resolved.text;
    },

    tr(value, { params = {}, defaultText = null, forceKey = false, returnMeta = false } = {}) {
        if (value === null || value === undefined) {
            return returnMeta ? { text: '', meta: null } : '';
        }

        const rawValue = String(value);
        let keyUsed = null;
        let resolved = null;

        const shouldTreatAsKey = forceKey || (this._isKeyLike(rawValue) && (this.hasKey(rawValue) || this.hasKey(rawValue, 'en')));

        if (shouldTreatAsKey) {
            keyUsed = rawValue;
            resolved = this._resolveKey(rawValue, params);
        }

        if (!resolved || !resolved.found) {
            const text = defaultText !== null
                ? defaultText
                : (keyUsed ? this.missingTranslationPlaceholder : rawValue);
            const meta = {
                keyUsed,
                resolvedText: text,
                locale: this.currentLocale,
                fallbackUsed: false,
                missing: Boolean(keyUsed)
            };
            return returnMeta ? { text, meta } : text;
        }

        const meta = {
            keyUsed,
            resolvedText: resolved.text,
            locale: resolved.localeUsed,
            fallbackUsed: resolved.fallbackUsed,
            missing: false
        };

        return returnMeta ? { text: resolved.text, meta } : resolved.text;
    },
    
    /**
     * Get list of available locales.
     * 
     * @returns {Array<string>} Array of locale codes
     */
    getAvailableLocales() {
        return Object.keys(this.translations);
    },
    
    /**
     * Check if a locale is available.
     * 
     * @param {string} locale - Locale code to check
     * @returns {boolean} True if available
     */
    isLocaleAvailable(locale) {
        return locale in this.translations;
    }
};

// Auto-initialize with English on load (async)
(async () => {
    try {
        if (i18n.initialized || i18n.initInProgress) {
            return;
        }
        await i18n.init('en');
    } catch (error) {
        console.error('[i18n] Auto-initialization failed:', error);
    }
})();

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = i18n;
}

File: campaign_service_record/utils/i18n.py
"""
IL-2 Campaign Tracker - Internationalization (i18n) Helper

Provides translation functionality for all Python components:
- GUI (Tkinter)
- CLI messages
- Event generation
- PDF generation
- In-game text (info.locale files)

Usage:
    from campaign_service_record.utils.i18n import t, set_locale, get_locale
    
    # Set locale (once at startup)
    set_locale('de')  # or 'en', etc.
    
    # Translate strings
    t('ui.button.ok')  # → "OK" (en) or "OK" (de)
    t('ui.popup.promotion_message', rank='Hauptmann')  
    # → "Sie wurden zum Hauptmann befördert"
    
    # Get current locale
    current = get_locale()  # → 'de'
"""

import json
import logging
from pathlib import Path
from typing import Dict, Any, Optional

from utils.supported_locales import (
    APP_TO_IL2_LOCALE,
    DEFAULT_LOCALE,
    IL2_TO_APP_LOCALE,
    get_locales_dir,
    get_supported_locales,
    normalize_locale,
)

logger = logging.getLogger(__name__)

# Global state
_current_locale: str = DEFAULT_LOCALE
_translations: Dict[str, Dict[str, Any]] = {}
_locales_dir: Optional[Path] = None
_initialized: bool = False


def _get_locales_dir() -> Path:
    """
    Get the locales directory path.
    
    Works for both script and PyInstaller EXE mode:
    - Script: PROJECT_ROOT/locales
    - EXE: Same directory as EXE (locales/ next to .exe)
    """
    global _locales_dir
    if _locales_dir is None:
        _locales_dir = get_locales_dir()
    return _locales_dir


def load_locale(locale: str) -> Dict[str, Any]:
    """
    Load translation file for given locale.
    
    Args:
        locale: Locale code ('en', 'de', etc.)
        
    Returns:
        Translation dictionary (empty dict if file not found)
    """
    locales_dir = _get_locales_dir()
    file_path = locales_dir / f'{locale}.json'
    
    if not file_path.exists():
        logger.warning(f"Translation file not found: {file_path}")
        return {}
    
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
            logger.debug(f"Loaded locale '{locale}' from {file_path}")
            return data
    except (OSError, json.JSONDecodeError) as e:
        logger.error(f"Error loading locale '{locale}': {e}")
        return {}


def init_i18n(locale: str = DEFAULT_LOCALE, force_reload: bool = False):
    """
    Initialize i18n system with given locale.
    
    This loads the fallback (en) and requested locale into memory.
    Should be called once at application startup.
    
    Args:
        locale: Locale code to set
        force_reload: Force reload even if already initialized
    """
    global _initialized, _current_locale, _translations

    locale = normalize_locale(locale)
    
    if _initialized and not force_reload:
        if locale and locale != _current_locale:
            set_locale(locale)
        logger.debug(f"i18n already initialized (locale={_current_locale})")
        return
    
    # Load fallback (English) first
    _translations[DEFAULT_LOCALE] = load_locale(DEFAULT_LOCALE)
    
    if not _translations[DEFAULT_LOCALE]:
        logger.error("CRITICAL: English fallback translations not loaded!")
    
    # Load requested locale (if different from fallback)
    if locale != DEFAULT_LOCALE:
        _translations[locale] = load_locale(locale)
        if not _translations[locale]:
            logger.warning(f"Requested locale '{locale}' not available, using 'en' fallback")
            locale = DEFAULT_LOCALE
    
    _current_locale = locale
    _initialized = True
    logger.info(f"i18n initialized with locale='{locale}'")


def set_locale(locale: str):
    """
    Set current locale.
    
    If the locale hasn't been loaded yet, this will load it.
    
    Args:
        locale: Locale code ('en', 'de', etc.)
    """
    global _current_locale, _translations
    
    if not locale:
        return

    locale = normalize_locale(locale)

    # Ensure fallback is always loaded
    if DEFAULT_LOCALE not in _translations:
        _translations[DEFAULT_LOCALE] = load_locale(DEFAULT_LOCALE)
    
    # Load requested locale if not already loaded
    if locale not in _translations:
        _translations[locale] = load_locale(locale)
        if not _translations[locale]:
            logger.warning(f"Locale '{locale}' not available, staying with '{_current_locale}'")
            return
    
    _current_locale = locale
    logger.debug(f"Locale set to '{locale}'")


def get_locale() -> str:
    """
    Get current locale code.
    
    Returns:
        Current locale code (e.g., 'en', 'de')
    """
    return _current_locale


def get_loaded_locales() -> list[str]:
    """
    Get list of locales currently loaded in memory.

    Returns:
        List of locale codes.
    """
    return sorted(_translations.keys())


def get_il2_locale_code() -> str:
    """
    Get IL-2 locale code for current app locale.
    
    Returns:
        IL-2 locale code (e.g., 'eng', 'ger')
    """
    return APP_TO_IL2_LOCALE.get(_current_locale, 'eng')


def _get_nested_value(data: Dict[str, Any], key_parts: list) -> Any:
    """
    Navigate nested dictionary using key parts.
    
    Args:
        data: Dictionary to navigate
        key_parts: List of key parts (e.g., ['ui', 'button', 'ok'])
        
    Returns:
        Value at nested location, or None if not found
    """
    current = data
    for part in key_parts:
        if isinstance(current, dict) and part in current:
            current = current[part]
        else:
            return None
    return current


def t(key: str, **params) -> str:
    """
    Translate key with optional parameters.
    
    Fallback chain: current_locale → 'en' → '[key]'
    
    Args:
        key: Translation key (dot-separated, e.g., 'ui.button.ok')
        **params: Named parameters for string formatting
        
    Returns:
        Translated string with parameters substituted
        
    Examples:
        >>> t('ui.button.ok')
        'OK'
        
        >>> set_locale('de')
        >>> t('ui.popup.promotion_message', rank='Hauptmann')
        'Sie wurden zum Hauptmann befördert'
        
        >>> t('nonexistent.key')
        '[nonexistent.key]'
    """
    # Ensure initialized
    if not _translations:
        init_i18n()
    
    # Split nested key
    key_parts = key.split('.')
    
    # Try current locale
    trans = _get_nested_value(_translations.get(_current_locale, {}), key_parts)
    
    # If not found, try fallback (en)
    if not isinstance(trans, str):
        trans = _get_nested_value(_translations.get(DEFAULT_LOCALE, {}), key_parts)
    
    # If still not found, return placeholder
    if not isinstance(trans, str):
        logger.debug(f"Translation key not found: {key}")
        return f'[{key}]'
    
    # Parameter substitution
    if params:
        try:
            return trans.format(**params)
        except KeyError as e:
            # Missing parameter - return with error indicator
            logger.warning(f"Missing parameter for key '{key}': {e}")
            return f'{trans} [MISSING: {e}]'
        except Exception as e:
            logger.error(f"Error formatting key '{key}': {e}")
            return trans
    
    return trans


def get_available_locales() -> list[str]:
    """
    Get list of available locale codes.
    
    Returns:
        List of locale codes (e.g., ['en', 'de'])
    """
    return get_supported_locales()


def validate_translations() -> Dict[str, list]:
    """
    Validate that all locales have the same keys as English.
    
    Returns:
        Dictionary mapping locale codes to lists of missing keys
    """
    en_keys = _get_all_keys(_translations.get(DEFAULT_LOCALE, {}))
    missing_by_locale = {}
    
    for locale in get_available_locales():
        if locale == DEFAULT_LOCALE:
            continue
        
        if locale not in _translations:
            _translations[locale] = load_locale(locale)
        
        locale_keys = _get_all_keys(_translations.get(locale, {}))
        missing = en_keys - locale_keys
        
        if missing:
            missing_by_locale[locale] = sorted(missing)
    
    return missing_by_locale


def _get_all_keys(data: Dict[str, Any], prefix: str = '') -> set:
    """
    Recursively extract all keys from nested dictionary.
    
    Args:
        data: Dictionary to extract keys from
        prefix: Current key prefix (for recursion)
        
    Returns:
        Set of all dot-separated keys
    """
    keys = set()
    
    for key, value in data.items():
        full_key = f'{prefix}.{key}' if prefix else key
        
        if isinstance(value, dict):
            keys.update(_get_all_keys(value, full_key))
        else:
            keys.add(full_key)
    
    return keys


# Auto-initialize with English on import
init_i18n(DEFAULT_LOCALE)
